### PR TITLE
Add `ChrF++`

### LINKED ISSUE #634: Add `ChrF++`
## 🚀 Feature

Add `ChrF++`, a text metric used for NMT evaluation.

**Sources:**
Paper - [CHRF++: words helping character n-grams](https://aclanthology.org/W17-4770.pdf)
[Repo](https://github.com/m-popovic/chrF) / metric is also implemented in [`sacrebleu`](https://github.com/mjpost/sacrebleu/blob/master/sacrebleu/metrics/chrf.py)

### Motivation

Character n-gram F-score (CHRF) is shown to correlate very well with human relative rankings of different machine translation outputs, especially for morphologically rich target languages. ([Popović, 2017](https://aclanthology.org/W17-4770.pdf))

### Additional context
I've already started working on this metric, I'll open a PR soon. :]

### PR DESCRIPTION
**Pull request by stancld**
## What does this PR do?

This PR adds a new text evaluation metric - `ChrF++`.

Fixes #634 

**TODO:** docs

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]
I cannot comment much on the implementation since I am not that much into NLP, but since you test that against another lib I assume it to be fine :)

Just a very minor question

File: torchmetrics/functional/text/chrf.py [hunk 1]
```diff
@@ -0,0 +1,717 @@
...
+    return defaultdict(zero_tensor)
+
+
+def _defaultdict_of_tensors_int_keys() -> Dict[int, Tensor]:
+    """A wrapper for creating `defaultdict` with key type of `int` and initialized with a zero tensor."""
+
+    def zero_tensor() -> Tensor:
+        return tensor(0.0)
+
+    return defaultdict(zero_tensor)
```
Comment (justusschock) [inline torchmetrics/functional/text/chrf.py:98]: why can't we use `defaultdict(lambda : tensor(0.0))` for that? This function is only meant for initialization right? Is it ever pickled?

**Review by justusschock — COMMENTED** [review]
The missing question :D

File: docs/source/references/functional.rst [hunk 2]
```diff
@@ -482,3 +482,9 @@ word_information_preserved [func]
 
 .. autofunction:: torchmetrics.functional.word_information_preserved
     :noindex:
+
+chrf_score [func]
```
Comment (SkafteNicki) [inline docs/source/references/functional.rst:486]: please insert to be in alphabetical order

File: docs/source/references/modules.rst [hunk 3]
```diff
@@ -665,6 +665,12 @@ WordInfoPreserved
 .. autoclass:: torchmetrics.WordInfoPreserved
     :noindex:
 
+CHRFScore
```
Comment (SkafteNicki) [inline docs/source/references/modules.rst:668]: please insert to be in alphabetical order

File: torchmetrics/text/chrf.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        dist_sync_fn: Optional[Callable] = None,
+    ):
+        super().__init__(
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn,
+        )
+
+        self.n_char_order = n_char_order
```
Comment (SkafteNicki) [inline torchmetrics/text/chrf.py:113]: lets add input validation :]
```suggestion
        if not isinstance(n_char_order, int) or n_char_order < 1:
            raise ValueError("Expected argument `n_char_order` to be an integer higher than 0")
        self.n_char_order = n_char_order
```

File: torchmetrics/text/chrf.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        for (n_gram_level, n_gram_order), text in self._get_text_n_gram_iterator():
+            for n in range(1, n_gram_order + 1):
+                dict_name = self._get_dict_name(text, n_gram_level)
+                state_name = self._get_state_name(text, n_gram_level, n)
+
+                n_grams_dicts[dict_name][n] = getattr(self, state_name)
+
+        return tuple(n_grams_dicts.values())  # type: ignore
+
+    def _update_states_from_dicts(self, n_grams_dicts_tuple: _DICT_STATES_TYPES) -> None:
```
Comment (SkafteNicki) [inline torchmetrics/text/chrf.py:189]:
```suggestion
    def _update_states_from_dicts(self, n_grams_dicts_tuple: _DICT_STATES_TYPES) -> None:
        """ Update global metric states based on the n grad dict calculated on the current batch """
```

File: torchmetrics/text/chrf.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+            (Optionally) A list of sentence-level chrF/chrF++ scores if `return_sentence_level_score=True`.
+        """
+        if self.sentence_chrf_score is not None:
+            return (
+                _chrf_score_compute(*self._convert_states_to_dicts(), self.n_order, self.beta),
+                torch.cat(self.sentence_chrf_score),
+            )
+        return _chrf_score_compute(*self._convert_states_to_dicts(), self.n_order, self.beta)
+
+    def _convert_states_to_dicts(self) -> _DICT_STATES_TYPES:
```
Comment (SkafteNicki) [inline torchmetrics/text/chrf.py:172]: please add docstring

File: torchmetrics/text/chrf.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+        Return:
+            A corpus-level chrF/chrF++ score.
+            (Optionally) A list of sentence-level chrF/chrF++ scores if `return_sentence_level_score=True`.
+        """
+        if self.sentence_chrf_score is not None:
+            return (
+                _chrf_score_compute(*self._convert_states_to_dicts(), self.n_order, self.beta),
+                torch.cat(self.sentence_chrf_score),
+            )
+        return _chrf_score_compute(*self._convert_states_to_dicts(), self.n_order, self.beta)
```
Comment (SkafteNicki) [inline torchmetrics/text/chrf.py:170]: nit:
```suggestion
        return _chrf_score_compute(
          *self._convert_states_to_dicts(), 
          self.n_order, 
          self.beta,
          torch.cat(self.sentence_chrf_score) if self.sentence_chrf_score is not None else None,
        )     
```

File: torchmetrics/functional/text/chrf.py (hunk 1 above)
Comment (stancld) [reply torchmetrics/functional/text/chrf.py]: Yes, that's a definitely good idea to use lambda functions instead! :]

**Review by SkafteNicki — APPROVED** [review]

File: torchmetrics/functional/text/chrf.py [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+) -> Dict[int, Tensor]:
+    """Get a number of n-gram matches between reference and hypothesis n-grams.
+
+    Args:
+        ref_n_grams_counts:
+        ref_n_grams_counts:
+
+    Return:
+    """
+    matching_n_grams = defaultdict(lambda: tensor(0.0))
```
Comment (SkafteNicki) [inline torchmetrics/functional/text/chrf.py:221]: `from typing import DefaultDict` needed at the top
```suggestion
    matching_n_grams: DefaultDict = defaultdict(lambda: tensor(0.0))
```

File: torchmetrics/text/chrf.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    def _get_dict_name(text: str, n_gram_level: str) -> str:
+        """Return a dictionary name w.r.t input args."""
+        return f"total_{text}_{n_gram_level}_n_grams"
+
+    @staticmethod
+    def _get_state_name(text: str, n_gram_level: str, n: int) -> str:
+        """Return a metric state name w.r.t input args."""
+        return f"total_{text}_{n_gram_level}_{n}_grams"
+
+    def _get_text_n_gram_iterator(self) -> itertools.product:
```
Comment (SkafteNicki) [inline torchmetrics/text/chrf.py:224]:
```suggestion
    def _get_text_n_gram_iterator(self) -> Iterator:
```

**Review by SkafteNicki — COMMENTED** [review]
only some typing issues remain :]
